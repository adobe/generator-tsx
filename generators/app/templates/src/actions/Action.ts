import { Action as ReduxAction } from 'redux'

export default interface Action<T> extends ReduxAction<T> {
	type: T
}
