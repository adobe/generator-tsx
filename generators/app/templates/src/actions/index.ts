import { GlobalActions } from './global/globalActionTypes'

export default RootActions

// The following type must remain at the bottom of this file in order for the
// generator to properly append new actions to it.
// prettier-ignore
type RootActions =
	| { type: 'REMOVE_ME' }
	| GlobalActions
