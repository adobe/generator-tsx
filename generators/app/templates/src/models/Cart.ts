import CartItem from './CartItem'
import Currency from './Currency'

export default interface Cart {
	items: CartItem[]
	subtotal: Currency
}
