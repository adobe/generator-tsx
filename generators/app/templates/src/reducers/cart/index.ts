export * from './cartReducer'
export { default } from './cartReducer'
