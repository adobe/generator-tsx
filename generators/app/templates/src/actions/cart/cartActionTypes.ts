import Cart from 'models/Cart'

import Action from '../Action'

export const FETCH_CART_REQUEST = 'FETCH_CART_REQUEST'
export interface FetchCartRequest extends Action<typeof FETCH_CART_REQUEST> {}

export const FETCH_CART_SUCCESS = 'FETCH_CART_SUCCESS'
export interface FetchCartSuccess extends Action<typeof FETCH_CART_SUCCESS> {
	cart: Cart
}

export const FETCH_CART_FAILURE = 'FETCH_CART_FAILURE'
export interface FetchCartFailure extends Action<typeof FETCH_CART_FAILURE> {
	errors: Error[]
}

export type CartActions = FetchCartRequest | FetchCartSuccess | FetchCartFailure
