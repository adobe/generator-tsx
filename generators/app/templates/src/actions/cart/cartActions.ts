import * as cartAPI from 'api/cart'
import Cart from 'models/Cart'
import ResponseError from 'models/ResponseError'

import ThunkAction from '../ThunkAction'

import * as types from './cartActionTypes'

export function fetchCart(): ThunkAction<void> {
	return async dispatch => {
		dispatch(fetchCartRequest())
		try {
			dispatch(fetchCartSuccess(await cartAPI.fetchCart()))
		} catch (errors) {
			dispatch(fetchCartFailure(errors))
		}
	}

	function fetchCartRequest(): types.FetchCartRequest {
		return { type: types.FETCH_CART_REQUEST }
	}

	function fetchCartSuccess(cart: Cart): types.FetchCartSuccess {
		return { type: types.FETCH_CART_SUCCESS, cart }
	}

	function fetchCartFailure(errors: ResponseError[]): types.FetchCartFailure {
		return { type: types.FETCH_CART_FAILURE, errors }
	}
}
