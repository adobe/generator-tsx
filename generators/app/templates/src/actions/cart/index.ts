import * as types from './cartActionTypes'

export * from './cartActions'
export { types }
