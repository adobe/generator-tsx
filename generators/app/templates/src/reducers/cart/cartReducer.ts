import update from 'immutability-helper'

import { types } from 'actions/cart'
import createReducer from 'helpers/createReducer'
import Cart from 'models/Cart'

export interface CartState extends Cart {
	errors: Error[]
	loading: boolean
}

const defaultState: CartState = {
	errors: [],
	items: [],
	loading: false,
	subtotal: {
		amount: 0,
	},
}

export default createReducer('cart', defaultState)(
	(state, action: types.CartActions) => {
		switch (action.type) {
			case types.FETCH_CART_FAILURE:
				return update(defaultState, {
					errors: { $set: action.errors },
				})
			case types.FETCH_CART_REQUEST:
				return update(state, {
					loading: { $set: true },
				})
			case types.FETCH_CART_SUCCESS:
				return update(defaultState, {
					$merge: action.cart,
				})
			default:
				return state
		}
	},
)
