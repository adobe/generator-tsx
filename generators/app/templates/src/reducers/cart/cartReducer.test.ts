import { types } from 'actions/cart'
import Cart from 'models/Cart'
import testReducer from 'utils/test/testReducer'

import cartReducer from './cartReducer'

testReducer(cartReducer, test => {
	test(
		{ loading: false },
		{ type: types.FETCH_CART_REQUEST },
		{ loading: true },
	)

	{
		const newCart: Cart = {
			items: [],
			subtotal: {
				amount: 42,
			},
		}
		test(
			{
				items: [],
				subtotal: {
					amount: 0,
				},
			},
			{
				type: types.FETCH_CART_SUCCESS,
				cart: newCart,
			},
			newCart,
		)
	}

	{
		const newErrors = [new Error('test')]
		test(
			{ errors: [] },
			{
				type: types.FETCH_CART_FAILURE,
				errors: newErrors,
			},
			{ errors: newErrors },
		)
	}
})
