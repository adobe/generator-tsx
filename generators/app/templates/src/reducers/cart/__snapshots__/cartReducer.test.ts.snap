// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartReducer returns initial state 1`] = `
Object {
  "errors": Array [],
  "items": Array [],
  "loading": false,
  "subtotal": Object {
    "amount": 0,
  },
}
`;
